File: maze.py
import settings
import box
import random



#==============Functions==============#

#Create a new Maze and return its map.
def newMaze() :
    createdMaze = Maze()
    createdMaze.build()
    createdMaze.Map()
    return createdMaze.map

#==============Class==============#

#a maze is made of boxes. Two of them are special : an exit, and an enter.


class Maze:
    def __init__(self):
        self.core = [] #the core of the maze contains al the boxes at their right location
        self.map = [] #map of the maze (boxes stored as bytes, see box.py for details)
        self.boxesToChange = [] #list of boxes that are not yet finished, using their coords
        

        #initialization of 2D lists with empty boxes
        for row in range(settings.ROWS) :
            self.core.append([])
            self.map.append([])
            for _column in range(settings.COLUMNS) :
                self.core[row].append(box.Box())
                self.map[row].append([])

        #initialization of their values
        column = 0
        row = 0
        for cpt in range(settings.ROWS*settings.COLUMNS) :
            self.core[row][column].groupNumber = cpt
            self.core[row][column].coordX = row
            self.core[row][column].coordY = column

            self.boxesToChange.append(((row,column))) #Will be useful for building the maze
            column += 1

            if (column)==settings.COLUMNS : #when we are at the end of the line
                column=0
                row += 1
        
        self.core[0][0].makeEnter()
        self.core[settings.ROWS-1][settings.COLUMNS-1].makeExit()
        self.boxesToChange.remove((0,0))


    #This function is used next. 
    #It updates all the boxes groupnumber depending on the groupnumber of the chosen box, and its adjacent box.
    #See Maze.Build() for the choice of the box and its adjacent box. .

    def updateGN(self,adjacentGN,currentGN) :
        if adjacentGN < currentGN:
                    for coordX in range(settings.ROWS) :
                        for coordY in range(settings.COLUMNS) :
                            if self.core[coordX][coordY].groupNumber == currentGN :
                                self.core[coordX][coordY].groupNumber = adjacentGN
        else:
            for coordX in range(settings.ROWS) :
                for coordY in range(settings.COLUMNS) :
                    if self.core[coordX][coordY].groupNumber == adjacentGN :
                        self.core[coordX][coordY].groupNumber = currentGN
        

    """functions for case...switch like statement used in Maze.build()
    They chose the wall to destroy, following two rules :
    -A wall on the edge of a maze cannot be destroyed
    -A wall separating two boxes with the same group number cannot be destroyed
    The three functions are very similar
    """

    def north(self,x,y) :
        if (x!=0) : #first rule
            adjacentBoxGN = self.core[x-1][y].groupNumber
            if adjacentBoxGN != self.core[x][y].groupNumber: #second rule
                self.core[x][y].removeWallNorth()
                self.core[x-1][y].removeWallSouth()
                self.updateGN(adjacentBoxGN,self.core[x][y].groupNumber)
                
                
        if self.core[x][y].groupNumber == 0 : #if our box is in the same group as the enter
            self.boxesToChange.remove((x,y)) # we remove it from the boxesToChange list, this box is done !

    def east(self,x,y) :
        if (y!=settings.COLUMNS-1) : #first rule
            adjacentBoxGN = self.core[x][y+1].groupNumber
            if adjacentBoxGN != self.core[x][y].groupNumber: #second rule
                self.core[x][y].removeWallEast()
                self.core[x][y+1].removeWallWest()
                self.updateGN(adjacentBoxGN,self.core[x][y].groupNumber)
                
        if self.core[x][y].groupNumber == 0 : #if our box is in the same group as the enter
            self.boxesToChange.remove((x,y)) # we remove it from the boxesToChange list, this box is done !

    def south(self,x,y) :
        if (x!=settings.ROWS-1) : #first rule
            adjacentBoxGN = self.core[x+1][y].groupNumber
            if adjacentBoxGN != self.core[x][y].groupNumber: #second rule
                self.core[x][y].removeWallSouth()
                self.core[x+1][y].removeWallNorth()
                self.updateGN(adjacentBoxGN,self.core[x][y].groupNumber)
                
        if self.core[x][y].groupNumber == 0 : #if our box is in the same group as the enter
            self.boxesToChange.remove((x,y)) # we remove it from the boxesToChange list, this box is done !

    def west(self,x,y) :
        if (y!=0) : #first rule
            adjacentBoxGN = self.core[x][y-1].groupNumber
            if adjacentBoxGN != self.core[x][y].groupNumber: #second rule
                self.core[x][y].removeWallWest()
                self.core[x][y-1].removeWallEast()
                self.updateGN(adjacentBoxGN,self.core[x][y].groupNumber)
                
        if self.core[x][y].groupNumber == 0 : #if our box is in the same group as the enter
            self.boxesToChange.remove((x,y)) # we remove it from the boxesToChange list, this box is done !
    
    

    #builds the maze, using the 4 previous functions
    def build(self) :
        while (len(self.boxesToChange) != 0):
            x,y = random.choice(self.boxesToChange)
            randomWall = random.randint(1,4)
            _directions = {1:self.north , 2:self.east , 3:self.south , 4:self.west ,}[randomWall](x,y)


    def Map(self):
        column = 0
        row = 0
        for _cpt in range(settings.ROWS*settings.COLUMNS) :
            self.map[row][column] = self.core[row][column].walls
            column += 1
            if (column)==settings.COLUMNS : #when we are at the end of the line
                column=0
                row += 1




        

File: maze.cpp
//cell.cpp
//Creation : 21/03/20

#include "maze.hpp"
#include <random>
#include <thread>

maze::maze(int x, int y)
{
    MAX_X = x;
    MAX_Y = y;
    //Core and map initialization + Initialization of the array of coords of the cells to modify
    mazeCore.resize(MAX_X);
    mazeMap.resize(MAX_X);
    for (int x = 0; x < MAX_X; x++)
    {
        mazeMap[x].resize(MAX_Y);
        for (int y = 0; y < MAX_Y; y++)
        {
            mazeCore[x].emplace_back(x + y * MAX_X + 1, x, y);
            unreachableCoords.emplace(x, y);
        }
    }

    mazeCore[0][0].makeEnter();
    unreachableCoords.erase({0, 0});
    mazeCore[MAX_X - 1][MAX_Y - 1].makeExit();
}

maze::~maze() {}

std::vector<std::vector<uint8_t>> maze::get_mazeMap() { return mazeMap; }
std::vector<std::vector<cell>> maze::get_mazeCore() { return mazeCore; }
int maze::get_MAX_X() { return MAX_X; }
int maze::get_MAX_Y() { return MAX_Y; }
void maze::set_MAX_X(int x) { MAX_X = x; }
void maze::set_MAX_Y(int y) { MAX_Y = y; }

void maze::removeSurrounded()
{
    std::set<std::pair<int, int>> surroundedCoords;
    while (unreachableCoords.size() > 100)
    {
        for (auto coordsToCheck : unreachableCoords)
        {
            int x = coordsToCheck.first;
            int y = coordsToCheck.second;
            bool surrounded = true;

            if (y != 0) //Rule #1
            {
                if (mazeCore[x][y - 1].get_groupNumber() != mazeCore[x][y].get_groupNumber()) //Rule #2
                {
                    surrounded = false;
                }
            }
            if ((surrounded) and (x != MAX_X - 1)) //Rule #1
            {
                if (mazeCore[x + 1][y].get_groupNumber() != mazeCore[x][y].get_groupNumber()) //Rule #2
                {
                    surrounded = false;
                }
            }
            if ((surrounded) and (y != MAX_Y - 1)) //Rule #1
            {
                if (mazeCore[x][y + 1].get_groupNumber() != mazeCore[x][y].get_groupNumber()) //Rule #2
                {
                    surrounded = false;
                }
            }
            if ((surrounded) and (x != 0)) //Rule #1
            {
                if (mazeCore[x - 1][y].get_groupNumber() != mazeCore[x][y].get_groupNumber()) //Rule #2
                {
                    surrounded = false;
                }
            }
            if (surrounded)
            {
                surroundedCoords.insert(coordsToCheck);
            }
        }

        if (surroundedCoords.size() > 1)
        {
            unreachableMutex.lock();
            for (auto &&coord : surroundedCoords)
            {
                unreachableCoords.erase(coord);
            }
            unreachableMutex.unlock();
            surroundedCoords.clear();
        }
    }
}

//Update the maze groupnumbers following build's rules (see maze::build())
void maze::updateGroups(int adjacentGN, int currentGN)
{
    int GNtoApply, GNtoModify;
    if (adjacentGN < currentGN)
    {
        GNtoApply = adjacentGN;
        GNtoModify = currentGN;
    }
    else
    {
        GNtoApply = currentGN;
        GNtoModify = adjacentGN;
    }
    for (int x = 0; x < MAX_X; x++)
    {
        for (int y = 0; y < MAX_Y; y++)
        {
            if (mazeCore[x][y].get_groupNumber() == GNtoModify)
            {
                mazeCore[x][y].set_groupNumber(GNtoApply);
                if (GNtoApply == 1) //If the cell is reachable from the enter
                {
                    unreachableCoords.erase({x, y});
                }
            }
        }
    }
}

/*
-Cells that are accessible from the enter are in the group 1
-We choose a random wall from a cell which isn't in group 1, and destroy a random adjacent wall following building rules
-We update the GroupNumbers using the previous function :
    - The modified cell is the chosen one. The adjacent is the one sharing the destroyed wall with the modified one.
    - If the modified cell has a smaller GN than the adjacent one, modified cell GN is modified
    - Else, the adjacent cell and all cells in its group get the modified cell's group number
-When all cells are in group 1, the build is done, we draw and return the map */
void maze::build()
{
    std::thread t1(&maze::removeSurrounded, this);

    //Random init
    std::random_device rd;
    std::mt19937 rng(rd());
    std::uniform_int_distribution<int> uni14(1, 4);

    int cpt = 0;

    while (unreachableCoords.size() > 0)
    {
        unreachableMutex.lock();
        //std::cout << unreachableCoords.size() << std::endl;
        //We choose a random cell and a random direction for the wall to destroy
        std::uniform_int_distribution<int> uniCells(0, unreachableCoords.size() - 1);
        int randomNumber = uniCells(rng);
        auto iter = unreachableCoords.begin();
        advance(iter, randomNumber);
        int x = (*iter).first;
        int y = (*iter).second;
        int chosenWall = uni14(rng);

        int test = 1;
        while (test < 5)
        {
            switch (chosenWall)
            {
            case 1:         //North
                if (y != 0) //Rule #1
                {
                    if (mazeCore[x][y - 1].get_groupNumber() != mazeCore[x][y].get_groupNumber()) //Rule #2
                    {
                        mazeCore[x][y].removeWallNorth();
                        mazeCore[x][y - 1].removeWallSouth();
                        updateGroups(mazeCore[x][y - 1].get_groupNumber(), mazeCore[x][y].get_groupNumber());
                        test = 5;
                    }
                }
                break;
            case 2:                 //East
                if (x != MAX_X - 1) //Rule #1
                {
                    if (mazeCore[x + 1][y].get_groupNumber() != mazeCore[x][y].get_groupNumber()) //Rule #2
                    {
                        mazeCore[x][y].removeWallEast();
                        mazeCore[x + 1][y].removeWallWest();
                        updateGroups(mazeCore[x + 1][y].get_groupNumber(), mazeCore[x][y].get_groupNumber());
                        test = 5;
                    }
                }
                break;
            case 3:                 //South
                if (y != MAX_Y - 1) //Rule #1
                {
                    if (mazeCore[x][y + 1].get_groupNumber() != mazeCore[x][y].get_groupNumber()) //Rule #2
                    {
                        mazeCore[x][y].removeWallSouth();
                        mazeCore[x][y + 1].removeWallNorth();
                        updateGroups(mazeCore[x][y + 1].get_groupNumber(), mazeCore[x][y].get_groupNumber());
                        test = 5;
                    }
                }
                break;
            case 4:         //West
                if (x != 0) //Rule #1
                {
                    if (mazeCore[x - 1][y].get_groupNumber() != mazeCore[x][y].get_groupNumber()) //Rule #2
                    {
                        mazeCore[x][y].removeWallWest();
                        mazeCore[x - 1][y].removeWallEast();
                        updateGroups(mazeCore[x - 1][y].get_groupNumber(), mazeCore[x][y].get_groupNumber());
                        test = 5;
                    }
                }
                break;
            }
            chosenWall = (chosenWall + 1) % 5; //circular way to test next wall if it failed
            test++;
        }
        cpt++;
        unreachableMutex.unlock();
    }

    t1.join();

    //Now that the maze is built, we draw the map
    for (unsigned x = 0; x < mazeCore.size(); x++)
    {
        for (unsigned y = 0; y < mazeCore[x].size(); y++)
        {
            mazeMap[x][y] = mazeCore[x][y].get_walls();
        }
    }
}
